internal/venice-client-common/src/main/java/com/linkedin/venice/serializer/AvroSerializer.java

```diff
@@ -77,94 +59,26 @@ protected AvroSerializer(DatumWriter<K> genericDatumWriter, DatumWriter<K> speci
     this.buffered = buffered;
   }
 
-  private void write(K object, Encoder encoder) throws IOException {
-    try {
-      if (object instanceof SpecificRecord) {
-        specificDatumWriter.write(object, encoder);
-      } else {
-        genericDatumWriter.write(object, encoder);
-      }
-    } catch (NullPointerException e) {
-      if (object instanceof SpecificRecord && specificDatumWriter == null) {
-        /**
-         * Defensive code...
-         *
-         * At the time of writing this commit, only the {@link VsonAvroGenericSerializer}
-         * uses the protected constructor to pass in a null {@link specificDatumWriter},
-         * and the Vson serializer should never be used with a SpecificRecord, so the NPE
-         * should never happen. If this assumption is broken in the future, and this code
-         * regresses, then hopefully this exception can help future maintainers to find
-         * the issue more easily.
-         */
-        throw new IllegalStateException(
-            "This instance of " + this.getClass().getSimpleName()
-                + " was instantiated with a null specificDatumWriter, and was used to serialize a SpecificRecord.",
-            e);
-      }
-      throw e;
-    }
-  }
-
   @Override
   public byte[] serialize(K object) throws VeniceException {
-    ByteArrayOutputStream output = new ByteArrayOutputStream();
-    try {
-      return serialize(object, null, output);
-    } finally {
-      try {
-        output.close();
-      } catch (IOException e) {
-        LOGGER.error("Failed to close stream", e);
-      }
-    }
-  }
-
-  @Override
-  public byte[] serialize(K object, ReusableObjects reuse) throws VeniceException {
-    return serialize(object, reuse.getBinaryEncoder(), reuse.getByteArrayOutputStream());
-  }
-
-  @Override
-  public byte[] serialize(K object, BinaryEncoder reusedEncoder, ByteArrayOutputStream reusedOutputStream)
-      throws VeniceException {
-    reusedOutputStream.reset();
-    Encoder encoder = AvroCompatibilityHelper.newBinaryEncoder(reusedOutputStream, buffered, reusedEncoder);
+    ReusableObjects reusableObjects = REUSABLE_OBJECTS.get();
```

Comment:
Hmm, is this better? It looks simpler, for sure, but it means that in cases where we process N records in a loop, we'll lookup the thread local O(N) times rather than O(1). Maybe the thread local lookup is completely free, but I am not sure that is a safe assumption... 🤔

If it is a goal to squash three functions into one, then can we consider keeping just the one that accepts the reusable objects wrapper as a 2nd parameter, and get rid of the other two? This way the looping use cases could still achieve the O(1) access to the thread local?

Reply:
The primary goal is to ensure serialization utility objects aren't allocated on each request. It can be done without changing internal apis, but I think that the optimization was premature and broke encapsulation in several components, leaking serializer implementation details to even external users.

AFAIK, the only place that can benefit from thread-local access amortization is DVC batch-get. At the same time, I expect the difference is negligible, because DVC single-get is already really fast despite thread-local access on each request. Do you think JMH benchmark is required to clear this question?

Reply:
What do you mean by external users? Where are these APIs used externally?

We could do a JMH benchmark, sure...

Reply:
According to JMH benchmark, ThreadLocal usage is pretty cheap:
`Benchmark                                 Mode      Cnt      Score   Error  Units
ThreadLocalBenchmark.load               sample  9657110     39.343 ± 0.161  ns/op
ThreadLocalBenchmark.load:load·p0.00    sample              23.000          ns/op
ThreadLocalBenchmark.load:load·p0.50    sample              36.000          ns/op
ThreadLocalBenchmark.load:load·p0.90    sample              48.000          ns/op
ThreadLocalBenchmark.load:load·p0.95    sample              50.000          ns/op
ThreadLocalBenchmark.load:load·p0.99    sample              55.000          ns/op
ThreadLocalBenchmark.load:load·p0.999   sample              63.000          ns/op
ThreadLocalBenchmark.load:load·p0.9999  sample           11012.622          ns/op
ThreadLocalBenchmark.load:load·p1.00    sample           89472.000          ns/op
ThreadLocalBenchmark.noop               sample  6926962     34.183 ± 0.219  ns/op
ThreadLocalBenchmark.noop:noop·p0.00    sample              20.000          ns/op
ThreadLocalBenchmark.noop:noop·p0.50    sample              31.000          ns/op
ThreadLocalBenchmark.noop:noop·p0.90    sample              37.000          ns/op
ThreadLocalBenchmark.noop:noop·p0.95    sample              43.000          ns/op
ThreadLocalBenchmark.noop:noop·p0.99    sample              50.000          ns/op
ThreadLocalBenchmark.noop:noop·p0.999   sample              58.000          ns/op
ThreadLocalBenchmark.noop:noop·p0.9999  sample           12304.000          ns/op
ThreadLocalBenchmark.noop:noop·p1.00    sample           39232.000          ns/op`